File: LeetCode/77_Combinations.py
# https://leetcode.com/problems/combinations/
# 2022年07月11日 14:18:26

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        # sub-problem of permutation
        # v1 copied 
        # https://leetcode.com/problems/combinations/discuss/26990/Python-easy-to-understand-DFS-solution
        ret=[]
        nums=list(range(1,n+1))
        def recursion(nums,k,path,ret):
            if len(path)==k:
                ret.append(path)
                return
            for i in range(len(nums)):
                recursion(nums[i+1:],k,path+[nums[i]],ret)
        recursion(nums,k,[],ret)
        return ret
        

# iterative backtracking
# https://leetcode.com/problems/combinations/discuss/27029/AC-Python-backtracking-iterative-solution-60-ms

File: LeetCode/77_Combinations.go
// https://leetcode.com/problems/combinations/
// 2023年03月27日 13:45:51

func combine(n int, k int) [][]int {
    nums:=[]int{}
    for i:=1;i<=n;i++{
        nums=append(nums,i)
    }
    ret:=[][]int{}
    bt(nums,k,[]int{},&ret)
    return ret
}

func bt(n []int, k int, path []int, ret *[][]int){
    if len(path)==k{
        newPath:=make([]int,len(path))
        copy(newPath,path) // 每次均开辟新地址，防止ret里的所有值都指向同一个path
        *ret=append(*ret,newPath)
    }
    for idx,val:=range n{
        fmt.Println(val)
        fmt.Println(n[idx+1:])
        path=append(path,val)
        bt(n[idx+1:],k,path,ret)
        path=path[:len(path)-1]
    }
}